source/agora/test/EnrollmentManager.d

```diff
@@ -221,3 +221,142 @@ unittest
+        {
+            this.ledger = ledger;
+            super(network, key_pair, ledger, taskman);
+        }
+
+        ///
+        public override bool isNominating () @safe @nogc nothrow
+        {
+            return this.is_nominating;
+        }
+
+        ///
+        public override void tryNominate () @safe
+        {
+            // Checks done in `Nominator.tryNominate`
+            if (this.is_nominating)
+                return;
+
+            this.is_nominating = true;
+            scope (exit) this.is_nominating = false;
+
+            ConsensusData data;
+            this.ledger.prepareNominatingSet(data);
+            if (data.tx_set.length == 0)
+                return;  // not ready yet
+
+            // Add an enrollment that is already an validator
+            data.enrolls ~= cast(Enrollment)BadNominator.dup_en;
+
+            // note: we are not passing the previous tx set as we don't really
+            // need it at this point (might later be necessary for chain upgrades)
+            auto slot_idx = this.ledger.getBlockHeight() + 1;
+            this.nominate(slot_idx, data);
+        }
+    }
+
+    static class MisbehavingValidator : TestValidatorNode
+    {
+        /// Ctor
+        public this (Config config, Registry* reg, immutable(Block)[] blocks)
+        {
+            super(config, reg, blocks);
+        }
+
+        ///
+        protected override Nominator getNominator (
+            NetworkManager network, KeyPair key_pair, Ledger ledger,
+            TaskManager taskman)
+        {
+            return new BadNominator(network, key_pair, ledger, taskman);
+        }
+    }
+
+    static class BadAPIManager : TestAPIManager
+    {
+        ///
+        public this (immutable(Block)[] blocks)
+        {
+            super(blocks);
+        }
+
+        /// see base class
+        public override void createNewNode (Config conf)
+        {
+            if (this.nodes.length == 1)
+            {
+                auto api = RemoteAPI!TestAPI.spawn!MisbehavingValidator(
+                    conf, &this.reg, this.blocks, conf.node.timeout);
+                this.reg.register(conf.node.address, api.tid());
+                this.nodes ~= NodePair(conf.node.address, api);
+            }
+            else
+            {
+                super.createNewNode(conf);
+            }
+        }
+    }
+
+    auto network = makeTestNetwork!BadAPIManager(TestConf.init);
+    scope(exit) network.shutdown();
+    scope(failure) network.printLogs();
+    network.start();
+    network.waitForDiscovery();
+    auto nodes = network.clients;
+    auto validator = nodes[0];
+    auto bad_validator = nodes[1];
+
+    // Sanity check: Check if genesis block has enrollments
+    const b0 = validator.getBlocksFrom(0, 2)[0];
+    assert(b0.header.enrollments.length >= 1);
+
+    // Set an duplicated enrollment to `BadNominator`
+    BadNominator.dup_en = cast(shared)b0.header.enrollments[0];
```

Comment:
I'm not sure if the BadNominator was created.

Reply:
The `BadNominator` is created in the `createNewNode` of the `BadAPIManager` when the `makeTestNetwork`  is called. The code you point out is to set the duplicated enrollment to the static variable. Thanks.